Mathlib/Algebra/Order/Incidence.lean

```diff
@@ -0,0 +1,645 @@
+  · exact h _ _ hab
+  · rw [apply_eq_zero_of_not_le hab, apply_eq_zero_of_not_le hab]
+
+@[simp] lemma mk_coe (f : IncidenceAlgebra 𝕜 α) (h) : mk f h = f := rfl
+
+end Coes
+
+/-! ### Additive and multiplicative structure -/
+
+instance instZero : Zero (IncidenceAlgebra 𝕜 α) := ⟨⟨fun _ _ ↦ 0, fun _ _ _ ↦ rfl⟩⟩
+instance instInhabited : Inhabited (IncidenceAlgebra 𝕜 α) := ⟨0⟩
+
+@[simp, norm_cast] lemma coe_zero : ⇑(0 : IncidenceAlgebra 𝕜 α) = 0 := rfl
+lemma zero_apply (a b : α) : (0 : IncidenceAlgebra 𝕜 α) a b = 0 := rfl
+
+end Zero
+
+section Add
+variable [AddZeroClass 𝕜] [LE α]
+
+instance instAdd : Add (IncidenceAlgebra 𝕜 α) where
+  add f g := ⟨f + g, fun a b h ↦ by simp_rw [Pi.add_apply, apply_eq_zero_of_not_le h, zero_add]⟩
+
+@[simp, norm_cast] lemma coe_add (f g : IncidenceAlgebra 𝕜 α) : ⇑(f + g) = f + g := rfl
+lemma add_apply (f g : IncidenceAlgebra 𝕜 α) (a b : α) : (f + g) a b = f a b + g a b := rfl
+
+end Add
+
+section Smul
+variable {M : Type*} [Zero 𝕜] [LE α] [SMulZeroClass M 𝕜]
+
+instance instSmulZeroClassRight : SMulZeroClass M (IncidenceAlgebra 𝕜 α) where
+  smul c f :=
+    ⟨c • ⇑f, fun a b hab ↦ by simp_rw [Pi.smul_apply, apply_eq_zero_of_not_le hab, smul_zero]⟩
+  smul_zero c := by ext; exact smul_zero _
+
+@[simp, norm_cast] lemma coe_constSMul (c : M) (f : IncidenceAlgebra 𝕜 α) : ⇑(c • f) = c • ⇑f := rfl
+
+lemma constSMul_apply (c : M) (f : IncidenceAlgebra 𝕜 α) (a b : α) : (c • f) a b = c • f a b := rfl
+
+end Smul
+
+instance instAddMonoid [AddMonoid 𝕜] [LE α] : AddMonoid (IncidenceAlgebra 𝕜 α) :=
+  DFunLike.coe_injective.addMonoid _ coe_zero coe_add fun _ _ ↦ rfl
+
+instance instAddCommMonoid [AddCommMonoid 𝕜] [LE α] : AddCommMonoid (IncidenceAlgebra 𝕜 α) :=
+  DFunLike.coe_injective.addCommMonoid _ coe_zero coe_add fun _ _ ↦ rfl
+
+section AddGroup
+variable [AddGroup 𝕜] [LE α]
+
+instance instNeg : Neg (IncidenceAlgebra 𝕜 α) where
+  neg f := ⟨-f, fun a b h ↦ by simp_rw [Pi.neg_apply, apply_eq_zero_of_not_le h, neg_zero]⟩
+
+instance instSub : Sub (IncidenceAlgebra 𝕜 α) where
+  sub f g := ⟨f - g, fun a b h ↦ by simp_rw [Pi.sub_apply, apply_eq_zero_of_not_le h, sub_zero]⟩
+
+@[simp, norm_cast] lemma coe_neg (f : IncidenceAlgebra 𝕜 α) : ⇑(-f) = -f := rfl
+@[simp, norm_cast] lemma coe_sub (f g : IncidenceAlgebra 𝕜 α) : ⇑(f - g) = f - g := rfl
+lemma neg_apply (f : IncidenceAlgebra 𝕜 α) (a b : α) : (-f) a b = -f a b := rfl
+lemma sub_apply (f g : IncidenceAlgebra 𝕜 α) (a b : α) : (f - g) a b = f a b - g a b := rfl
+
+instance instAddGroup : AddGroup (IncidenceAlgebra 𝕜 α) :=
+  DFunLike.coe_injective.addGroup _ coe_zero coe_add coe_neg coe_sub (fun _ _ ↦ rfl) fun _ _ ↦ rfl
+
+end AddGroup
+
+instance instAddCommGroup [AddCommGroup 𝕜] [LE α] : AddCommGroup (IncidenceAlgebra 𝕜 α) :=
+  DFunLike.coe_injective.addCommGroup _ coe_zero coe_add coe_neg coe_sub (fun _ _ ↦ rfl)
+    fun _ _ ↦ rfl
+
+section One
+variable [Preorder α] [DecidableEq α] [Zero 𝕜] [One 𝕜]
+
+instance instOne : One (IncidenceAlgebra 𝕜 α) :=
```

Comment:
We should add a comment somewhere that `1` refers to the delta function, and mention "delta" by name